irohad/ametsuchi/impl/tx_presence_cache_impl.cpp

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ametsuchi/impl/tx_presence_cache_impl.hpp"
+#include "common/visitor.hpp"
+#include "interfaces/iroha_internal/transaction_batch.hpp"
+#include "interfaces/transaction.hpp"
+
+namespace iroha {
+  namespace ametsuchi {
+    TxPresenceCacheImpl::TxPresenceCacheImpl(std::shared_ptr<Storage> storage)
+        : storage_(storage) {}
```

Comment:
`std::move(storage)`?

Reply:
Please, no. Performance optimisation for one-time action doesn't worth all debug time losses if storage creator decides to use this storage after move.

Reply:
But you pass a copy here, so it's impossible to use it after move since the creator will have a different instance

Reply:
I mean a next scenario:
```
std::shared_ptr<Storage> storage = std::make_shared<...>(...);
auto cache = TxPresenceCacheImpl(storage);
....
storage-> ...   <=== trouble here
```

Reply:
not really since you pass it by value. the original storage will remain intact

Reply:
Nickita is right, the pointer, which cam to this ctor, is already a copy of the original one

Reply:
True, agree with you guys.